sql/sql_connect.cc

```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
   }
 }
 
+void reset_global_user_or_client_stats(HASH* stats)
+{
+  DYNAMIC_ARRAY delete_items;
+
+  if (my_init_dynamic_array(&delete_items, sizeof(USER_STATS *),
+        stats->records, 1))
+  {
+    sql_print_error("reset global stats: cannot initialize dynamic array");
+    my_message(ER_OUTOFMEMORY, ER(ER_OUTOFMEMORY), MYF(0));
+    return;
+  }
+
+  for (uint idx=0;idx < stats->records; idx++)
+  {
+    USER_STATS* user_stats=
+      (USER_STATS *) my_hash_element(stats, idx);
+    if (user_stats->concurrent_connections)
+    {
+      init_user_stats(user_stats, NULL, NULL,
+          0, 0, user_stats->concurrent_connections, // connections
+          0, 0, 0,   // time
+          0, 0, 0,   // bytes sent, received and written
+          0, 0, 0,   // rows fetched, updated and read
+          0, 0, 0,   // select, update and other commands
+          0, 0,      // commit and rollback trans
+          0,         // diff_denied_connections
+          0,         // lost connections
+          0,         // access denied errors
+          0);        // empty queries
+    }
+    else
+    {
+      if (insert_dynamic(&delete_items, (uchar *) &user_stats))
+      {
+        delete_dynamic(&delete_items);
+        sql_print_error("reset global stats: insert_dynamic failed");
+        my_message(ER_OUTOFMEMORY, ER(ER_OUTOFMEMORY), MYF(0));
+        return;
+      }
+    }
+  }
+
+  for (uint i= 0; i < delete_items.elements; ++i)
+  {
+    USER_STATS *entry= *dynamic_element(&delete_items, i, USER_STATS **);
+    (void) my_hash_delete(stats, (uchar*) entry);
```

Comment:
It's a bug if this my_hash_delete fails - please DBUG_ASSERT its return value; C++-style cast